lib/auth/grpcserver_test.go

```diff
@@ -1046,33 +1034,62 @@ func TestIsMFARequired(t *testing.T) {
 			Hostname: "node-a",
 		},
 	}
-	_, err = srv.Auth().UpsertNode(ctx, node)
+	_, err := srv.Auth().UpsertNode(ctx, node)
 	require.NoError(t, err)
 
 	// Create a fake user.
 	user, role, err := CreateUserAndRole(srv.Auth(), "no-mfa-user", []string{"no-mfa-user"})
 	require.NoError(t, err)
 
-	for _, required := range []bool{true, false} {
-		t.Run(fmt.Sprintf("required=%v", required), func(t *testing.T) {
-			roleOpt := role.GetOptions()
-			roleOpt.RequireSessionMFA = required
-			role.SetOptions(roleOpt)
-			err = srv.Auth().UpsertRole(ctx, role)
-			require.NoError(t, err)
+	for _, authPrefRequireMFAType := range []types.RequireMFAType{
+		types.RequireMFAType_OFF,
+		types.RequireMFAType_SESSION,
+		types.RequireMFAType_SESSION_AND_HARDWARE_KEY,
+		types.RequireMFAType_HARDWARE_KEY_TOUCH,
+	} {
+		authPref, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+			Type:           constants.Local,
+			SecondFactor:   constants.SecondFactorOptional,
+			RequireMFAType: authPrefRequireMFAType,
+			Webauthn: &types.Webauthn{
+				RPID: "teleport",
+			},
+		})
+		require.NoError(t, err)
+		err = srv.Auth().SetAuthPreference(ctx, authPref)
+		require.NoError(t, err)
 
-			cl, err := srv.NewClient(TestUser(user.GetName()))
-			require.NoError(t, err)
+		for _, roleRequireMFAType := range []types.RequireMFAType{
+			types.RequireMFAType_OFF,
+			types.RequireMFAType_SESSION,
+			types.RequireMFAType_SESSION_AND_HARDWARE_KEY,
+			types.RequireMFAType_HARDWARE_KEY_TOUCH,
+		} {
+			// If role or auth pref have "hardware_key_touch", expect not required.
+			expectRequired := !(roleRequireMFAType == types.RequireMFAType_HARDWARE_KEY_TOUCH || authPrefRequireMFAType == types.RequireMFAType_HARDWARE_KEY_TOUCH)
+			// Otherwise, if auth pref or role require session MFA, expect required.
+			expectRequired = expectRequired && (roleRequireMFAType.IsSessionMFARequired() || authPrefRequireMFAType.IsSessionMFARequired())
+
+			t.Run(fmt.Sprintf("authPref=%v/role=%v/expect=%v", authPrefRequireMFAType, roleRequireMFAType, expectRequired), func(t *testing.T) {
+				roleOpt := role.GetOptions()
```

Comment:
It takes ~2.5s for me locally. Since it's changing the test server's Role and AuthPref for each subtest, it can't be run in parallel. 

I tested recreating the server for each subtest in order to parallize the tests, but it results in the same ~2.5s run while using more computing resources.